File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        
        ListNode temp = head;
        HashSet<ListNode> set = new HashSet<>();
        
        while(temp!=null){
            if(!set.contains(temp)){
                set.add(temp);
                temp = temp.next;
            }else{
                return temp;
            }
        }
        return null;
    }
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        unordered_set<ListNode*> st;
        ListNode *temp = head;
        while(temp){
            if(st.count(temp)){
                return temp;
            }else{
                st.insert(temp);
            }
            temp = temp->next;
        }
        return NULL;
    }
};